### PR TITLE
MoE FFN Last Block: tandem-specialized feed-forward expert in final TransolverBlock

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Round 26 confirmed that **slice routing is NOT the tandem failure driver** (attn-logit-noise σ=0.05, PR #2263, was clearly negative: p_in +3.9%). The model routes tandem samples correctly — the failure is in what the backbone **processes**, not how it routes.

Every major p_tan win has come from giving the model more explicit inter-foil signal as **input features** (gap/stagger bias, TE frame, wake deficit), not from attention routing changes. This points to a processing bottleneck: the shared FFN in the last TransolverBlock must produce a single feature manifold that generalizes across two physically distinct regimes (single-foil potential flow vs. tandem wake-affected flow).

**Hypothesis:** Giving tandem samples a **dedicated weight matrix** for the core nonlinear transformation in the last block will allow the backbone to assemble tandem-optimal representations without contaminating single-foil gradients.

**Why the last block only:** Conditioning all blocks failed (adaln_all_blocks). Earlier blocks learn shared geometry and aerodynamics primitives — only the last block needs to specialize. This is the minimal change that breaks the structural compromise at exactly the right layer.

**Why this is distinct from existing approaches:**
- `domain_velhead` / `soft_moe`: conditional OUTPUT HEADS (what is decoded), not FFN computation
- `domain_layernorm` / `adaln_output`: conditional LayerNorm affine transforms, not FFN weights
- `pressure_first` / `pressure_deep`: output channel ordering, not FFN
- This is the **first experiment** to give tandem samples a dedicated FFN weight matrix in the backbone.

### 1. Add config flag

In `Config` dataclass:
```python
moe_ffn: bool = False  # Tandem-specialized FFN expert in last TransolverBlock
```

Add `--moe_ffn` to argparse.

### 2. Modify TransolverBlock.__init__

Add `moe_ffn=False` parameter. When enabled, create a second FFN:
```python
if moe_ffn:
    self.mlp_tandem = MLP(hidden_dim, hidden_dim * mlp_ratio * 4, hidden_dim, n_layers=0, res=False)
self.moe_ffn = moe_ffn
```

The existing `self.mlp` becomes the single-foil expert. Use the exact same MLP constructor as the existing `self.mlp`.

### 3. Modify TransolverBlock.forward

Replace the FFN line:
```python
fx = _ln(self.ln_2_post, self.mlp(_ln(self.ln_2, fx)) + fx)
```

With:
```python
if self.moe_ffn:
    fx_ln = _ln(self.ln_2, fx)
    is_tan = (tandem_mask.view(-1) > 0.5)  # [B] boolean
    out = self.mlp(fx_ln)                   # default: single-foil expert
    if is_tan.any():
        tan_idx = is_tan.nonzero(as_tuple=True)[0]
        out[tan_idx] = self.mlp_tandem(fx_ln[tan_idx])
    fx = _ln(self.ln_2_post, out + fx)
else:
    fx = _ln(self.ln_2_post, self.mlp(_ln(self.ln_2, fx)) + fx)
```

**Critical:** `tandem_mask` must be available in the block's forward call. Check how `domain_velhead` or `soft_moe` accesses the tandem flag — follow the exact same pattern for passing it through.

### 4. Apply only to last block

In `Transolver.__init__`, when constructing the TransolverBlock list:
```python
TransolverBlock(..., moe_ffn=(self.moe_ffn if idx == n_layers - 1 else False))
```

Same pattern as `soft_moe`, `domain_velhead`, `pressure_first` — check how those flags are passed in the constructor loop and mirror it exactly.

### 5. Run configuration (2 seeds)

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent alphonse \
  --wandb_name "alphonse/moe-ffn-last-block-s42" \
  --wandb_group "moe-ffn-last-block" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --moe_ffn \
  --seed 42

# Seed 73: same flags with --seed 73 --wandb_name "alphonse/moe-ffn-last-block-s73"
```

Only change vs baseline: `--moe_ffn`.

### 6. What to report

Table: p_in, p_oodc, p_tan, p_re for each seed and 2-seed avg vs baseline.
W&B run IDs for both seeds.
Key signals: **p_tan is the primary target.** p_in should NOT regress (single-foil path is unchanged). If p_in regresses, the tandem_mask dispatch is incorrect — check the implementation.

If promising: suggest follow-up of applying moe_ffn to all 3 blocks (not just last) as a variation.

## Baseline (PR #2251, 2-seed avg)

| Metric | Value | Target to beat |
|--------|-------|----------------|
| p_in   | 11.891 | < 11.89 |
| p_oodc | 7.561  | < 7.56  |
| p_tan  | 28.118 | < 28.12 |
| p_re   | 6.364  | < 6.36  |

Baseline W&B runs: `7jix2jkg` (s42), `epkfhxfl` (s73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent alphonse --wandb_name "alphonse/baseline-tmax150" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results

### Summary Table

| Metric | Baseline (avg) | s42 (06hnldau) | s73 (8bsvpfuc) | Avg | Delta |
|--------|---------------|----------------|----------------|-----|-------|
| p_in   | 11.891 | 12.811 | 12.238 | **12.524** | +0.633 (+5.3%) ❌ |
| p_oodc | 7.561  | 8.053  | 8.054  | **8.054**  | +0.493 (+6.5%) ❌ |
| p_tan  | 28.118 | 29.076 | 28.993 | **29.034** | +0.916 (+3.3%) ❌ |
| p_re   | 6.364  | 6.787  | 6.889  | **6.838**  | +0.474 (+7.4%) ❌ |

All splits regressed vs baseline. **Negative result.**

### W&B Run IDs

- Seed 42: `06hnldau` (alphonse/moe-ffn-last-block-s42, 141 epochs)
- Seed 73: `8bsvpfuc` (alphonse/moe-ffn-last-block-s73, 142 epochs)
- W&B group: `moe-ffn-last-block`

### Commands used

```bash
# Seed 42 — CUDA_VISIBLE_DEVICES=0
cd cfd_tandemfoil && python train.py --agent alphonse \
  --wandb_name "alphonse/moe-ffn-last-block-s42" --wandb_group "moe-ffn-last-block" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --moe_ffn --seed 42
# Seed 73: same flags with --seed 73, CUDA_VISIBLE_DEVICES=1
```

### What happened

**p_in regressed significantly (+5.3%)**, which the PR noted as a signal of incorrect dispatch. However, I believe the dispatch itself is correct — the issue is the training dynamics of MoE with hard dispatch via `torch.where`.

**Root cause analysis:**

With hard dispatch (`torch.where(is_tan_mask, out_tandem, out_single)`), the gradient correctly flows only to the appropriate expert:
- `self.mlp` (single-foil expert) trains ONLY on non-tandem samples
- `self.mlp_tandem` trains ONLY on tandem samples

This halves the effective training data for each expert. The single-foil expert was previously trained on 100% of samples (the shared FFN saw everything); now it sees only ~70% of samples (non-tandem). This directly degrades single-foil p_in.

Note: I used `torch.where` instead of the PR's in-place `out[tan_idx] = ...` because in-place modification of an autograd tensor causes version counter issues with torch.compile. The correct gradient behavior of `torch.where` means the two experts are properly separated, but at the cost of less data per expert.

**Why p_tan also regressed:** The tandem expert starts from random init, has fewer training samples per epoch than the original shared FFN, and 142 epochs may not be enough to converge a brand-new expert. The shared FFN had the advantage of having learned useful representations from all samples.

**The p_in regression is not a dispatch bug** — it's a fundamental challenge with hard MoE: data efficiency drops for each expert.

### Suggested follow-ups

1. **Soft dispatch with learned mixing weight**: Replace hard masking with a learned scalar `alpha = sigmoid(gate(fx_ln))` applied per-sample. This lets both experts train on all samples while still allowing specialization. Formula: `out = alpha * out_t + (1-alpha) * out_s` where `alpha` is conditioned on tandem_mask.

2. **Residual tandem correction** (lighter alternative): Keep `self.mlp` for all samples, add a **residual correction** `self.mlp_tandem_delta` that is ADDED for tandem samples only: `out = self.mlp(fx_ln) + is_tan[:, None, None] * self.mlp_tandem_delta(fx_ln)`. This preserves the single-foil expert's full-data training, and the tandem delta trains only on tandem samples for specialization. No routing information loss.

3. **More data for tandem**: Increase tandem augmentation to give the tandem expert more gradient signal per epoch to compensate for the data reduction.

Option 2 (residual delta) is likely the strongest follow-up — it eliminates the data-efficiency problem entirely while still allowing tandem specialization.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2268 because ADVISOR: Closing — all 4 metrics regressed vs baseline (p_in +5.3%, p_oodc +6.5%, p_tan +3.3%, p_re +7.4%).

**Root cause:** Hard MoE dispatch with 2 experts halves the effective training data per expert, starving both. The tandem-specialized expert sees only ~50% of samples, which is insufficient for the data-hungry Transolver backbone. The gating network also adds optimization noise early in training.

**Takeaway:** MoE-style routing at the FFN level is incompatible with our small dataset regime. Expert specialization requires either (a) much more data, or (b) soft mixture routing (but that converges to a weighted average, defeating the purpose). Added to DO NOT REVISIT.

Good observation about residual delta approach — noted for future reference, but the fundamental data-per-expert problem remains..
